training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala

```diff
@@ -289,6 +317,64 @@ object DecisionTreeTrainer {
     }
   }
 
+  def giniImpurity(dist : scala.collection.mutable.Map[String, Double]) : Double = {
+    val sum = dist.map(x => x._2).sum
+    val scale = 1.0 / (sum * sum)
+    var impurity : Double = 0.0
+    for (kv1 <- dist) {
+      for (kv2 <- dist) {
+        if (kv1._1 != kv2._1) {
+          impurity += scale * kv1._2 * kv2._2
+        }
+      }
+    }
+    impurity
+  }
+
+  // Evaluate a multiclass classification-type split
+  def evaluateMulticlassSplit(examples : Array[util.Map[java.lang.String, util.Map[java.lang.String, java.lang.Double]]],
+                              rankKey : String,
+                              minLeafCount : Int,
+                              splitCriteria : String,
+                              candidateOpt : Option[ModelRecord]): Option[Double] = {
+    val leftDist = scala.collection.mutable.HashMap[String, Double]()
+    val rightDist = scala.collection.mutable.HashMap[String, Double]()
+
+    for (example <- examples) {
+      val response = BoostedStumpsModel.getStumpResponse(candidateOpt.get, example)
+      for (kv <- example.get(rankKey).asScala) {
+        if (response) {
+          rightDist.put(kv._1, kv._2)
```

Comment:
I feel the value should be accumulated, otherwise this will be determined by the last example only. how about something like?
`val v = rightDist.getOrElse(kv._1)`

`rightDist.put(kv._1, v + kv._2)`